alburnum/maas/viscera/boot_sources.py

```diff
@@ -0,0 +1,74 @@
+"""Objects for boot sources."""
+
+__all__ = [
+    "BootSource",
+    "BootSources",
+]
+
+from . import (
+    check,
+    check_optional,
+    Object,
+    ObjectField,
+    ObjectSet,
+    ObjectType,
+    parse_timestamp,
+)
+
+
+class BootSourcesType(ObjectType):
+    """Metaclass for `BootSources`."""
+
+    def __iter__(cls):
+        return map(cls._object, cls._handler.read())
+
+    def create(cls, url, *, keyring_filename=None, keyring_data=None):
+        """Create a new `BootSource`."""
+        if keyring_filename is None and keyring_data is None:
+            raise ValueError(
+                "Either keyring_filename and keyring_data must be set.")
+        data = cls._handler.create(
+            url=url,
+            keyring_filename=(
+                "" if keyring_filename is None else keyring_filename),
```

Comment:
Its the filename that is located on the MAAS server itself. It must be provided unless the url endswith ".json". I will fix it to check to see if ".json" is provided and then the ValueError will not be raised.